README.md
# semantic-release-jira

[**semantic-release**](https://github.com/semantic-release/semantic-release) plugin to publish a jira release.

![Github](https://github.com/dylanhitt/semantic-release-jira/actions/workflows/release.yml/badge.svg)

[![npm latest version](https://img.shields.io/npm/v/@temptek/semantic-release-jira/latest.svg)](https://www.npmjs.com/package/@temptek/semantic-release-jira)

| Step               | Description                                                                                                                                   |
|--------------------|----------------------------------------------------------------------------|
| `verifyConditions` | Validate the config options and check for a credentials in the environment |
| `sucess`           | Find all tickets from commits and add them to a new release on JIRA        |

## Install

```bash
$ npm install --save-dev @temptek/semantic-release-jira
$ yarn add --dev @temptek/semantic-release-jira
```

### Environment variables

| Variable                | Description                                                                                                                                                               |
| ----------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `JIRA_AUTH`             | Base64 encoded string of `username:password`.                                                                                                                             |
| `JIRA_USERNAME`         | Username in Jira. Used for basic auth.                                                                                                                                    |
| `JIRA_PASSWORD`         | Password in Jira. Used for basic auth.                                                                                                                                    |
| `JIRA_EMAIL`            | Email address in Jira. Used for basic auth.                                                                                                                               |
| `JIRA_API_TOKEN`        | Api token in Jira. Used for basic auth. [How to create an api token?](https://support.atlassian.com/atlassian-account/docs/manage-api-tokens-for-your-atlassian-account/) |          

Use either `JIRA_AUTH`, `JIRA_USERNAME` and `JIRA_PASSWORD` **or** `JIRA_EMAIL` and `JIRA_API_TOKEN`.

### Configuration

The plugin should be added to your config

```json
{
  "plugins": [
    "@semantic-release/commit-analyzer",
    "@semantic-release/release-notes-generator",
    "@semantic-release/git",
    [
      "@temptek/semantic-release-jira",
      {
        "projectId": "UH",
        "releaseNameTemplate": "Test v${version}",
        "jiraHost": "uphabit.atlassian.net",
        "ticketPrefixes": [
          "TEST",
          "UH"
        ],
        "ticketRegex": "[a-zA-Z]{3,5}-\\d{3,5}"
      }
    ]
  ]
}
```
Please note that `ticketRegex` cannot be used together with `ticketPrefixes`.

```typescript
interface Config {
    /**
     * A domain of a jira instance ie: `uphabit.atlasian.net`
     */
    jiraHost: string;

    /**
     * A list of prefixes to match when looking for tickets in commits. Cannot be used together with ticketRegex.
     *
     * ie. ['TEST'] would match `TEST-123` and `TEST-456`
     */
    ticketPrefixes?: string[];

    /**
     * A unescaped regex to match tickets in commits (without slashes). Cannot be used together with ticketPrefixes.
     *
     * ie. [a-zA-Z]{4}-\d{3,5} would match any ticket with 3 letters a dash and 3 to 5 numbers, such as `TEST-456`, `TEST-5643` and `TEST-56432`
     */
    ticketRegex?: string;

    /**
     * The id or key for the project releases will be created in
     */
    projectId: string;

    /**
     * A lodash template with a single `version` variable
     * defaults to `v${version}` which results in a version that is named like `v1.0.0`
     * ex: `Semantic Release v${version}` results in `Semantic Release v1.0.0`
     *
     * @default `v${version}`
     */
    releaseNameTemplate?: string;

    /**
     * A lodash template for the release.description field
     *
     * template variables:
     *    version: the sem-ver version ex.: 1.2.3
     *      notes: The full release notes: This may be very large
     *             Only use it if you have very small releases
     *
     * @default `Automated release with semantic-release-jira https://git.io/JvAbj`
     */
    releaseDescriptionTemplate?: string;

    /**
     * The number of maximum parallel network calls
     *
     * @default: 10
     */
    networkConcurrency?: number;

    /**
     * Indicates if a new release created in jira should be set as released
     */
    released?: boolean;

    /**
     * Include the release date when creating a release in jira
     *
     * @default: false
     */
    setReleaseDate?: boolean;

    /**
     * Specify board name to be used to find an active sprint.
     * Use the start date of the active sprint as start date for the release
     *
     * @default: undefined
     */
    useBoardForActiveSprint?: string;
}
```


CHANGELOG.md
## [1.2.1](https://github.com/dylanhitt/semantic-release-jira/compare/v1.2.0...v1.2.1) (2023-06-30)


### Bug Fixes

* verifyConditions was written by an idiot ([#7](https://github.com/dylanhitt/semantic-release-jira/issues/7)) ([ec4cb0a](https://github.com/dylanhitt/semantic-release-jira/commit/ec4cb0aef940ca27948a0c4e8b085bd1aed201d9))

# [1.2.0](https://github.com/dylanhitt/semantic-release-jira/compare/v1.1.0...v1.2.0) (2023-06-28)


### Features

* use eslint instead of eslint ([#6](https://github.com/dylanhitt/semantic-release-jira/issues/6)) ([517618a](https://github.com/dylanhitt/semantic-release-jira/commit/517618ae179230b90650be36d4a6d5040c833fda))

# [1.1.0](https://github.com/dylanhitt/semantic-release-jira/compare/v1.0.0...v1.1.0) (2023-04-28)


### Features

* let the user overwrite the released value ([6bcc749](https://github.com/dylanhitt/semantic-release-jira/commit/6bcc7498f7bb6178916a9a35091a63da95b5e7f2))

# 1.0.0 (2022-12-21)


### Bug Fixes

* [FIX-123] typescript config ([72159d8](https://github.com/dylanhitt/semantic-release-jira/commit/72159d8670218409327835f24226e814ea0559f9))
* add message when not run in prerelease branch ([ddeefc1](https://github.com/dylanhitt/semantic-release-jira/commit/ddeefc14ec6d641ee3d1d69c0bd19b3a2737c215))
* **ci:** Automate npm publish ([5abd275](https://github.com/dylanhitt/semantic-release-jira/commit/5abd2753f047122d1baaa4199acb97d5c3cfdb06))
* **ci:** make sure we can compile as a test ([05b883d](https://github.com/dylanhitt/semantic-release-jira/commit/05b883d5e18a804a276f80176709a73a05f20ff3))
* **ci:** make sure we can compile as a test ([542429b](https://github.com/dylanhitt/semantic-release-jira/commit/542429b40ce3e0dc96678f1ef3bb787472a6d91f))
* **ci:** remove extra unneeded dependency ([887f522](https://github.com/dylanhitt/semantic-release-jira/commit/887f52273a13e80f6bdc62b096572aa16b0fe6f7))
* **deps:** add missing tslib dep ([8b59b2c](https://github.com/dylanhitt/semantic-release-jira/commit/8b59b2c6468c37114a26d12f16f729ccf0e794c0))
* **deps:** upgrade jira-connector to 3.1.0 ([3d157d5](https://github.com/dylanhitt/semantic-release-jira/commit/3d157d587c368fc774e252f4aea9fd400083ad47))
* do not fail when board or sprint not found, log out the error though ([bbcc017](https://github.com/dylanhitt/semantic-release-jira/commit/bbcc017e4bba8b296d32821ad148069baa51e6a3))
* error when releaseDescriptionTemplate is undefined ([17dca0c](https://github.com/dylanhitt/semantic-release-jira/commit/17dca0cd685cb4c91859ff0af2943adb56108e00))
* escaping regex ([e1bea79](https://github.com/dylanhitt/semantic-release-jira/commit/e1bea7932f3c8ecd6647cadf6413500740137235))
* First version ([7b29389](https://github.com/dylanhitt/semantic-release-jira/commit/7b29389e260a0e1daaf2964ece5d694346a2d90d))
* First version ([bcca7ba](https://github.com/dylanhitt/semantic-release-jira/commit/bcca7bad30bf7d99c01cdb587bb4bf875690850c))
* First version ([30071f1](https://github.com/dylanhitt/semantic-release-jira/commit/30071f161520e055239740b2232e17c808b9b9b2))
* First version ([2fc24a7](https://github.com/dylanhitt/semantic-release-jira/commit/2fc24a758d22a305a60352ba3c25285e82a9efba))
* First version ([783d80f](https://github.com/dylanhitt/semantic-release-jira/commit/783d80f9953e98227dc0cdf5b24755a1b3217596))
* First version ([25ffcd6](https://github.com/dylanhitt/semantic-release-jira/commit/25ffcd68e0413d64cdf3f209a4718677415cab5f))
* First version ([f5557ff](https://github.com/dylanhitt/semantic-release-jira/commit/f5557ff7f571a34c3cedc9c360e34e1a579d72c1))
* First version ([a325b40](https://github.com/dylanhitt/semantic-release-jira/commit/a325b400d4f4d2f1e2a8416664025b0661345159))
* First version ([55491f6](https://github.com/dylanhitt/semantic-release-jira/commit/55491f631caed8b5656ddbd27e8ecb01b2cdaf7b))
* First version ([5276152](https://github.com/dylanhitt/semantic-release-jira/commit/52761521cd5b8ff7f860e20c33b46aa29c563c69))
* First version ([993ea98](https://github.com/dylanhitt/semantic-release-jira/commit/993ea98b9499c5484ba1637640fe539de177f2c1))
* First version ([dd714cf](https://github.com/dylanhitt/semantic-release-jira/commit/dd714cf64aad42a939bc1bdfc09b9ae85b4816a1))
* fixing ticketRegex matching ([7a3957d](https://github.com/dylanhitt/semantic-release-jira/commit/7a3957d5b14489ee7493f71f10f3e537f367b35e))
* Merge pull request [#20](https://github.com/dylanhitt/semantic-release-jira/issues/20) from UpHabit/renovate/major-semantic-release-monorepo ([d758907](https://github.com/dylanhitt/semantic-release-jira/commit/d758907d37bd0e861fd1b10ce3acbd6829006d4e))
* **security:** Merge pull request [#42](https://github.com/dylanhitt/semantic-release-jira/issues/42) from UpHabit/dependabot/npm_and_yarn/lodash-4.17.19 ([07a32d9](https://github.com/dylanhitt/semantic-release-jira/commit/07a32d93b56852e703c62cee5206fa05952bc9bc))
* **security:** update all packages with secuirty issues ([1ba6780](https://github.com/dylanhitt/semantic-release-jira/commit/1ba67803f6c1f51770bd0b8d0f0f1e23d9025e64))
* **version:** Release init version ([b7d5168](https://github.com/dylanhitt/semantic-release-jira/commit/b7d5168facc624f9808a61e9d0f4ed38687c5778))


### Features

* [UH-1258] better logging ([85a74e9](https://github.com/dylanhitt/semantic-release-jira/commit/85a74e937880e4cd5dd05dde47509db164855051))
* [UH-1258] Implement release creation ([c44e1aa](https://github.com/dylanhitt/semantic-release-jira/commit/c44e1aa0e8bab21f0aff2f00d4667625f8716c7d))
* add config for release description ([cd25e0d](https://github.com/dylanhitt/semantic-release-jira/commit/cd25e0d74174b4f2eff676cdf7dbb32e2e773f54))
* add option for released and releaseDate for veresion ([879d592](https://github.com/dylanhitt/semantic-release-jira/commit/879d592eae0f2ff7b321794fe7c1d386b8ce2dff))
* add option to not run on prerelease branches ([37fa1fc](https://github.com/dylanhitt/semantic-release-jira/commit/37fa1fc3194e3ac37186075623696f4fb092d659))

test/test.ts
import { getTickets } from '../lib/success';
import { PluginConfig } from '../lib/types';

import { context, pluginConfig } from './fakedata';

describe('Success tests', () => {
  describe('#getTickets', () => {
    it('should analyze tickets with one ticketPrefix', () => {
      const config = {
        ...pluginConfig,
        ticketPrefixes: ['UH'],
      } as PluginConfig;
      expect(getTickets(config, context)).toEqual(['UH-1258', 'UH-1234', 'UH-555']);
    });

    it('should get multiple tickets on the same commit', () => {
      const config = {
        ...pluginConfig,
        ticketPrefixes: ['TEST'],
      } as PluginConfig;
      expect(getTickets(config, context)).toEqual(['TEST-123', 'TEST-234']);
    });

    it('should analyze tickets with many ticketPrefix', () => {
      const config = {
        ...pluginConfig,
        ticketPrefixes: ['UH', 'FIX'],
      } as PluginConfig;
      expect(getTickets(config, context)).toEqual(['FIX-321', 'UH-1258', 'FIX-123', 'UH-1234', 'UH-555']);
    });

    it('should analyze tickets with ticketRegex', () => {
      const ticketRegex = '[A-Za-z]+-\\d+';

      const config: PluginConfig = {
        ...pluginConfig,
        ticketRegex,
      } as PluginConfig;

      expect(getTickets(config, context)).toEqual(['FIX-321', 'UH-1258', 'FIX-123', 'TEST-123', 'TEST-234', 'UH-1234', 'UH-555']);
    });
  });
});


test/fakedata.ts
import { BaseConfig, Commit, GenerateNotesContext, Person, PluginConfig, PluginContext, PreviousRelease, UpcomingRelease } from '../lib/types';

export const baseConfig: BaseConfig = {
  $0: '',
  branch: 'test',
  debug: true,
  dryRun: true,
};

export const date = new Date('2019-01-01T00:00:00.000Z');

export const author: Person = {
  date,
  name: 'test',
  email: 'email',
};

export const commits: Commit[] = [
  'chore: fixing whitespace',
  'docs: adding regex unescaping',
  'fix: escaping regex',
  'docs: [FIX-321] editing readme',
  'feat: [UH-1258] better logging ',
  'feat: [UH-1258] Implement release creation',
  'fix: [FIX-123] typescript config',
  'fix: [TEST-123] [TEST-234] test commit',
  'fix(UH-1234): This is a scoped commit',
  'feat(UH-555): This is another scoped commit',
].map(m => ({
  author,
  committer: author,
  commitDate: date,
  body: '',
  hash: '',
  message: m,
  subject: '',
  commit: {
    long: '',
    short: '',
  },
}));

export const previousRelease: PreviousRelease = {
  gitHead: '',
  gitTag: '',
  version: '',
};

export const upcomingRelease: UpcomingRelease = {
  ...previousRelease,
  notes: '',
  type: '',
};

export const pluginConfig: Partial<PluginConfig> = {
  ...baseConfig,
  projectId: 'TEST',
  jiraHost: 'testjira.com',
};

export const logger = {
  info: jest.fn(),
};

export const pluginContext: PluginContext = {
  cwd: '',
  env: {},
  logger: logger as any,
  options: baseConfig,
  stderr: null,
  stdout: null,
  branch: null,
};

export const context: GenerateNotesContext = {
  ...pluginContext,
  commits,
  lastRelease: previousRelease,
  nextRelease: upcomingRelease,
};


lib/jira.ts
import { AgileClient, Version3Client } from 'jira.js'
import type { Config } from 'jira.js'

import type { PluginConfig, PluginContext } from './types'

function getClientOptions (config: PluginConfig, context: PluginContext): Config {
  const clientOptions: Config = {
    host: config.jiraHost
  }

  if ((context.env.JIRA_USERNAME !== '') && (context.env.JIRA_PASSWORD !== '')) {
    clientOptions.authentication = {
      basic: {
        username: context.env.USERNAME,
        password: context.env.PASSWORD
      }
    }
  }

  if ((context.env.JIRA_EMAIL !== '') && (context.env.JIRA_API_TOKEN !== '')) {
    clientOptions.authentication = {
      basic: {
        email: context.env.JIRA_EMAIL,
        apiToken: context.env.JIRA_API_TOKEN
      }
    }
  }

  // This is for backwards compatibility
  if (context.env.JIRA_AUTH !== '') {
    const decoded = Buffer.from(context.env.JIRA_AUTH, 'base64').toString('utf-8')
    const username = decoded.split(':')[0]
    const password = decoded.split(':')[1]
    clientOptions.authentication = {
      basic: {
        username,
        password
      }
    }
  }

  return clientOptions
}

export function makeVersion3Client (config: PluginConfig, context: PluginContext): Version3Client {
  return new Version3Client(getClientOptions(config, context))
}

export function makeAgileClient (config: PluginConfig, context: PluginContext): AgileClient {
  return new AgileClient(getClientOptions(config, context))
}


lib/success.ts
import type { Version3, Version3Client } from 'jira.js'
import type { Sprint } from 'jira.js/out/agile/models'
import type { CreateVersion } from 'jira.js/out/version3/parameters'
import * as _ from 'lodash'
import pLimit from 'p-limit'

import { makeAgileClient, makeVersion3Client } from './jira'
import { DEFAULT_RELEASE_DESCRIPTION_TEMPLATE, DEFAULT_VERSION_TEMPLATE } from './types'
import type { GenerateNotesContext, PluginConfig } from './types'
import { escapeRegExp } from './util'

export function getTickets (config: PluginConfig, context: GenerateNotesContext): string[] {
  let patterns: RegExp[] = []

  if (config.ticketRegex != null) {
    patterns = [new RegExp(config.ticketRegex, 'giu')]
  } else if (config.ticketPrefixes != null) {
    patterns = config.ticketPrefixes
      .map(prefix => new RegExp(`\\b${escapeRegExp(prefix)}-(\\d+)\\b`, 'giu'))
  } else {
    context.logger.error('No config.ticketRegex or config.ticketPrefixes were provided, failed to find any tickets.')
    return []
  }

  const tickets = new Set<string>()
  for (const commit of context.commits) {
    for (const pattern of patterns) {
      const matches = commit.message.match(pattern)
      if (matches != null) {
        matches.forEach(match => {
          tickets.add(match)
          context.logger.info(`Found ticket ${matches.toString()} in commit: ${commit.commit.short}`)
        })
      }
    }
  }

  return [...tickets]
}

async function findOrCreateVersion (
  config: PluginConfig,
  context: GenerateNotesContext,
  jira: Version3Client,
  project: Version3.Version3Models.Project,
  name: string,
  description: string,
  activeSprint: Sprint | undefined
): Promise<Version3.Version3Models.Version> {
  const remoteVersions = project.versions
  context.logger.info(`Looking for version with name '${name}'`)
  const existing = _.find(remoteVersions, { name })
  if (existing?.id != null) {
    context.logger.info(`Found existing release '${existing.id}'`)
    return existing
  }

  context.logger.info('No existing release found, creating new')

  let newVersion: Version3.Version3Models.Version
  if (config.dryRun) {
    context.logger.info('dry-run: making a fake release')
    newVersion = {
      name,
      id: 'dry_run_id'
    } as any
  } else {
    const preRelease = typeof context.branch !== 'string' && Boolean(context.branch.prerelease)
    const parameters: CreateVersion = {
      name,
      description,
      projectId: project.id as any,
      startDate: activeSprint?.startDate,
      released: config.released ?? !preRelease,
      releaseDate: (config.setReleaseDate ?? false) ? new Date().toISOString() : undefined
    }
    newVersion = await jira.projectVersions.createVersion(parameters)
  }

  if (newVersion.id == null) {
    context.logger.error('Failed to create new version (couldn\'t find ID of new version)')
    return newVersion
  }

  context.logger.info(`Made new release '${newVersion.id}'`)
  return newVersion
}

async function editIssueFixVersions (config: PluginConfig, context: GenerateNotesContext, jira: Version3Client, newVersionNames: string[], releaseVersionIds: Array<string | undefined>, issueKey: string): Promise<void> {
  try {
    context.logger.info(`Adding issue ${issueKey} to '${newVersionNames.toString()}'`)
    if (!config.dryRun) {
      await jira.issues.editIssue({
        issueIdOrKey: issueKey,
        update: {
          fixVersions: releaseVersionIds.map(id => {
            return { add: { id } }
          })
        },
        properties: undefined as any
      })
    }
  } catch (err: any) {
    const allowedStatusCodes = [400, 404]
    let { statusCode }: { statusCode: number } = err
    if (typeof err === 'string') {
      try {
        const parsedErr = JSON.parse(err)
        statusCode = parsedErr.statusCode
      } catch (err) {
        // it's not json :shrug:
      }
    }
    if (!allowedStatusCodes.includes(statusCode)) {
      throw err
    }
    context.logger.error(`Unable to update issue ${issueKey} statusCode: ${statusCode}`)
  }
}

async function findActiveSprint (config: PluginConfig, context: GenerateNotesContext): Promise<Sprint | undefined> {
  if (config.useBoardForActiveSprint == null) {
    return undefined
  }

  const agileClient = makeAgileClient(config, context)
  const boards = await agileClient.board.getAllBoards({ projectKeyOrId: config.projectId })
  const board = boards.values.find(b => b.name === config.useBoardForActiveSprint)
  if (board != null) {
    const sprints = await agileClient.board.getAllSprints({ boardId: board.id })
    const activeSprint = sprints.values.find(s => s.state === 'active')
    return activeSprint
  }

  context.logger.error(`Board ${config.useBoardForActiveSprint} has no active sprint`)
  return undefined
}

function getVersionNames (config: PluginConfig, context: GenerateNotesContext): string[] {
  const templates = config.releaseNameTemplate == null ? [DEFAULT_VERSION_TEMPLATE] : (Array.isArray(config.releaseNameTemplate) ? config.releaseNameTemplate : [config.releaseNameTemplate])
  return templates.map(template => {
    // Parse the version into its components
    const [version, channel] = context.nextRelease.version.split('-')
    const [major, minor, patch] = version.split('.').map(Number)

    return _.template(template)({
      version: context.nextRelease.version,
      env: context.env,
      major,
      minor,
      patch,
      channel
    })
  })
}

export async function success (config: PluginConfig, context: GenerateNotesContext): Promise<void> {
  const isPrerelease = typeof context.branch !== 'string' && Boolean(context.branch.prerelease)
  const runOnPrerelease = config.runOnPrerelease === undefined || config.runOnPrerelease

  if (!isPrerelease || (isPrerelease && runOnPrerelease)) {
    const tickets = getTickets(config, context)

    context.logger.info(`Found ticket ${tickets.join(', ')}`)

    const versionNames = getVersionNames(config, context)

    const descriptionTemplate = _.template(config.releaseDescriptionTemplate ?? DEFAULT_RELEASE_DESCRIPTION_TEMPLATE)
    const newVersionDescription = descriptionTemplate({ version: context.nextRelease.version, notes: context.nextRelease.notes, env: context.env })

    context.logger.info(`Using jira release(s) '${versionNames.toString()}'`)

    const version3Client = makeVersion3Client(config, context)
    const project = await version3Client.projects.getProject({ projectIdOrKey: config.projectId })

    const activeSprint = await findActiveSprint(config, context)

    const concurrentLimit = pLimit(config.networkConcurrency ?? 10)

    const releaseVersionsPromises = versionNames.map(async (version: string) => {
      return await concurrentLimit(async () => await findOrCreateVersion(config, context, version3Client, project, version, newVersionDescription, activeSprint))
    })
    const releaseVersions = await Promise.all(releaseVersionsPromises)
    const releaseIds = releaseVersions.map(version => version.id)

    const edits = tickets.map(async issueKey => {
      await concurrentLimit(async () => { await editIssueFixVersions(config, context, version3Client, versionNames, releaseIds, issueKey) })
    })
    await Promise.all(edits)
  } else {
    context.logger.info('Configuration set to not run on prerelease branches')
  }
}


lib/verifyConditions.ts
/* eslint-disable @typescript-eslint/no-throw-literal */
import SemanticReleaseError from '@semantic-release/error'

import { makeVersion3Client } from './jira'
import type { PluginConfig, PluginContext } from './types'

export async function verifyConditions (config: PluginConfig, context: PluginContext): Promise<void> {
  const { networkConcurrency } = config

  if (typeof config.jiraHost !== 'string') {
    throw new SemanticReleaseError('config.jiraHost must be a string')
  }
  if (typeof config.projectId !== 'string') {
    throw new SemanticReleaseError('config.projectId must be a string')
  }

  if ((config.ticketPrefixes == null) && (config.ticketRegex == null)) {
    throw new SemanticReleaseError('Either config.ticketPrefixes or config.ticketRegex must be passed')
  }

  if ((config.ticketPrefixes != null) && (config.ticketRegex != null)) {
    throw new SemanticReleaseError('config.ticketPrefixes and config.ticketRegex cannot be passed at the same time')
  }

  if (config.ticketPrefixes != null) {
    if (!Array.isArray(config.ticketPrefixes)) {
      throw new SemanticReleaseError('config.ticketPrefixes must be an array of string')
    }
    for (const prefix of config.ticketPrefixes) {
      if (typeof prefix !== 'string') {
        throw new SemanticReleaseError('config.ticketPrefixes must be an array of string')
      }
    }
  }

  if ((config.ticketRegex != null) && typeof config.ticketRegex !== 'string') {
    throw new SemanticReleaseError('config.ticketRegex must be an string')
  }

  if (config.releaseNameTemplate != null) {
    if (typeof config.releaseNameTemplate !== 'string' && !Array.isArray(config.releaseNameTemplate)) {
      throw new SemanticReleaseError('config.releaseNameTemplate must be a string or a list of strings')
    }
  }

  if (config.releaseDescriptionTemplate != null) {
    if (typeof config.releaseDescriptionTemplate !== 'string') {
      throw new SemanticReleaseError('config.releaseDescriptionTemplate must be a string')
    }
  }

  if (config.useBoardForActiveSprint != null) {
    if (typeof config.useBoardForActiveSprint !== 'string') {
      throw new SemanticReleaseError('config.useBoardForActiveSprint must be a string')
    }
  }

  if ((networkConcurrency != null) && (typeof networkConcurrency !== 'number' || networkConcurrency < 1)) {
    throw new SemanticReleaseError('config.networkConcurrency must be an number greater than 0')
  }

  if ((context.env.JIRA_USERNAME === '') && (context.env.JIRA_PASSWORD === '') && (context.env.JIRA_EMAIL === '') && (context.env.JIRA_API_TOKEN === '') && (context.env.JIRA_AUTH === '')) {
    throw new SemanticReleaseError('Either JIRA_USERNAME and JIRA_PASSWORD or JIRA_EMAIL and JIRA_API_TOKEN must be set for basic auth')
  }

  if (((context.env.JIRA_USERNAME === '') && (context.env.JIRA_PASSWORD !== '')) || ((context.env.JIRA_USERNAME !== '') && (context.env.JIRA_PASSWORD === ''))) {
    throw new SemanticReleaseError('Both JIRA_USERNAME and JIRA_PASSWORD must be set for basic auth')
  }

  if (((context.env.JIRA_EMAIL === '') && (context.env.JIRA_API_TOKEN !== '')) || ((context.env.JIRA_EMAIL !== '') && (context.env.JIRA_API_TOKEN === ''))) {
    throw new SemanticReleaseError('Both JIRA_EMAIL and JIRA_API_TOKEN must be set for basic auth')
  }

  const jira = makeVersion3Client(config, context)

  await jira.projects.getProject({ projectIdOrKey: config.projectId })
}


lib/types.ts
import type { BranchSpec } from 'semantic-release'
import type { Signale } from 'signale'

export interface PluginContext {
  cwd: string
  env: Record<string, string>
  logger: Signale
  options: BaseConfig
  stderr: typeof process.stderr
  stdout: typeof process.stdout
  branch: BranchSpec
}

export interface Person {
  name: string
  email: string
  date: Date
}

export interface Commit {
  author: Person
  committer: Person
  commitDate: Date
  body: string
  hash: string
  message: string
  subject: string
  commit: {
    long: string
    short: string
  }
}

export interface PreviousRelease {
  gitHead: string
  gitTag: string
  version: string
}

export interface UpcomingRelease extends PreviousRelease {
  notes: string
  type: string
}

export interface GenerateNotesContext extends PluginContext {
  commits: Commit[]
  lastRelease: PreviousRelease
  nextRelease: UpcomingRelease
}

export interface BaseConfig {
  $0: string
  branch: string
  debug: boolean
  dryRun: boolean
}

// eslint-disable-next-line no-template-curly-in-string
export const DEFAULT_VERSION_TEMPLATE = 'v${version}'
export const DEFAULT_RELEASE_DESCRIPTION_TEMPLATE = 'Automated release with semantic-release-jira-releases https://git.io/JvAbj'

export interface PluginConfig extends BaseConfig {
  /**
   * A domain of a jira instance ie: `uphabit.atlasian.net`
   */
  jiraHost: string

  /**
   * A list of prefixes to match when looking for tickets in commits. Cannot be used together with ticketRegex.
   *
   * ie. ['TEST'] would match `TEST-123` and `TEST-456`
   */
  ticketPrefixes?: string[]

  /**
   * A unescaped regex to match tickets in commits (without slashes). Cannot be used together with ticketPrefixes.
   *
   * ie. [a-zA-Z]{4}-\d{3,5} would match any ticket with 3 letters a dash and 3 to 5 numbers, such as `TEST-456`, `TEST-5643` and `TEST-56432`
   */
  ticketRegex?: string

  /**
   * The id or key for the project releases will be created in
   */
  projectId: string

  /**
   * A lodash template with a single `version` variable
   * defaults to `v${version}` which results in a version that is named like `v1.0.0`
   * accepts a single template string or an array of template strings
   *
   * template variables:
   *    version: the sem-ver version ex.: 1.2.3
   *      major: the 1st segment of the semver
   *      minor: the 2nd segment of the semver
   *      patch: the 3rd segment of the semver
   *    channel: the optional "channel" semantic-release might add ex 1.2.3-prerelease.1 would have a channel of prerelease.1
   *
   * ex: `Semantic Release v${version}` results in `Semantic Release v1.0.0`
   *
   * @default `v${version}`
   */
  releaseNameTemplate?: string | string[]

  /**
   * A lodash template for the release.description field
   *
   * template variables:
   *    version: the sem-ver version ex.: 1.2.3
   *      notes: The full release notes: This may be very large
   *             Only use it if you have very small releases
   *
   * @default `Automated release with semantic-release-jira-releases https://git.io/JvAbj`
   */
  releaseDescriptionTemplate?: string

  /**
   * The number of maximum parallel network calls
   *
   * @default: 10
   */
  networkConcurrency?: number

  /**
   * Indicates if a new release created in jira should be set as released
   */
  released?: boolean

  /**
   * Include the release date when creating a release in jira
   *
   * @default: false
   */
  setReleaseDate?: boolean

  /**
   * Specify board name to be used to find an active sprint.
   * Use the start date of the active sprint as start date for the release
   *
   * @default: undefined
   */
  useBoardForActiveSprint?: string

  /**
   * Run on prerelease branch?
   *
   * @default: true
   */
  runOnPrerelease?: boolean
}
